### ISSUE TITLE
Q: any tricks to get it compiling on macOS/Darwin?

### ISSUE THREAD

**Issue by GwynethLlewelyn** [Tomas-M/iotop#41]
Hello there,

From the description of the project (I didn't delve deep into the code myself), it _seems_ that this tool (as opposed to the Python version) _assumes_ there is a Linux kernel running beneath the OS/POSIX layer. This may explain why no references are made to FreeBSD, OpenBSD, NetBSD etc. on the `README.md`.

I did a quick attempt to compile it under macOS/Darwin (which is vaguely BSD-ish); however, it seems that `clang` (the default compiler by Apple) and `gcc` 12.1 (installed through [Homebrew](https://brew.sh)) get utterly confused when trying to include something as simple as `src/iotop.h` — mostly because of the way `#define` is working:

```sh
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/types.h:75,
                 from src/iotop.h:27,
                 from src/arr.c:14:
/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/cdefs.h:681:49: error: operator '&&' has no right operand
  681 | #if defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE == 1L
      |                                                 ^~
/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/cdefs.h:687:49: error: operator '&&' has no right operand
  687 | #if defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE == 2L
      |                                                 ^~
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/cdefs.h:715:
/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/_posix_availability.h:32:79:error: operator '&&' has no right operand
   32 | #if !defined(_DARWIN_C_SOURCE) && defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 198808L
      |                                                                               ^~
/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/_posix_availability.h:38:79:error: operator '&&' has no right operand
   38 | #if !defined(_DARWIN_C_SOURCE) && defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 199009L
      |                                                                               ^~
/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/usr/include/sys/_posix_availability.h:44:79:error: operator '&&' has no right operand
   44 | #if !defined(_DARWIN_C_SOURCE) && defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 199209L
      |
[...]
```
My assumption is that the macOS/Darwin pre-processor has serious issues, after expanding the `#define`s, to deal with more complex expressions. This may perhaps be a result of an extra flag or a missing one; being an utter newbie to the complexities of compiling C/C++ (anything except the most basic things!), I wonder if anyone has a clue on what I ought to do to fix the above errors...

**Comment by bbonev** [comment]
Both `iotop-c` and `iotop-py` essentially rely on functionality specific to the Linux kernel. Thus running the current code on `*BSD` (that includes `MacOS`) is not possible.

In case someone contributes how to collect the IO data on some or all BSD flavors, I am open to add support for those platforms.

**Comment by GwynethLlewelyn** [comment]
@bbonev thank you for the clarification! Aye, I had assumed as much; also, I'm pretty sure that the Darwin kernel is not a 'pure' BSD Mach kernel, but may include lots of Apple changes, and therefore it might be hard to figure out how the IO data is _supposed_ to be extracted...
Granted, `iotop-py` _used_ to work until at least macOS Big Sur came out, and I'm pretty sure that the first versions of Big Sur still got it to run under **root**; however, recent upgrades have broken that functionality...

**Comment by bbonev** [comment]
> I'm pretty sure that the Darwin kernel is not a 'pure' BSD Mach kernel, but may include lots of Apple changes, and therefore it might be hard to figure out how the IO data is supposed to be extracted...

I'd wildly guess the API is similar to `FreeBSD` but I may be wrong.

> Granted, `iotop-py` used to work until at least macOS Big Sur came out

No way. Unless it was a different, `MacOS` specific implementation (`iotop-py` is located [here](https://repo.or.cz/w/iotop.git) and is Linux specific); I have seen somewhere mentions about a `MacOS` specific iotop implementation but do not remember URL and/or details...

To clarify - both `iotop-(py|c)` use the Linux specific netlink kernel API which is not present in any other OS.

**Comment by GwynethLlewelyn** [comment]
You're absolutely correct! What I've got is _not_ based on `iotop-py` at all; my mistake!

It all comes from having two completely different tools (but providing very similar output!), both named `iotop`, one of which I run on several Linux flavours (and architectures!) and one, well, that runs on my Mac. It never occurred to me before today that these might be totally unrelated — just sharing the same name. 

Indeed, instead of `iotop-(py|c)`, I have [a much, much older thing that uses `dtrace` to produce a decent amount of low-level tracing](https://www.brendangregg.com/dtrace.html), and which was originally developed for Solaris, but most BSD flavours ought to be able to run it — and macOS/Darwin did so, for a long while. [Linux support for `dtrace` is relatively more recent](https://www.brendangregg.com/blog/2016-10-27/dtrace-for-linux-2016.html), courtesy of [Oracle](https://github.com/oracle/dtrace-linux-kernel).

The original script I've got (well, more precisely, a dtrace programme), and that can be downloaded from Brendan Gregg's page is from... 2005! It has not been updated since 2006. Nevertheless, it worked quite well for me in the past 15 years or so. It was just when Apple finally managed to break it that I was looking for a replacement...

Thus, keep this thread closed :-) Again, my mistake, and I do apologise for the confusion I made.
